nunaliit2-couch-client/src/main/java/ca/carleton/gcrc/couch/client/impl/listener/AbstractCouchDbChangeListener.java

```diff
@@ -0,0 +1,137 @@
+ * queue in this class' thread, and subclass can do their work in {@link #processDocIdChanged(Pair)}.
+ */
+public abstract class AbstractCouchDbChangeListener extends Thread implements CouchDbChangeListener {
+    private static final Logger logger = LoggerFactory.getLogger(AbstractCouchDbChangeListener.class);
+    private static final long IDENTICAL_CHANGE_THRESHOLD_MILLIS = 5000L;
+    /**
+     * Indicates whether the thread is running.
+     */
+    private final AtomicBoolean running = new AtomicBoolean(false);
+    /**
+     * All change doc notifications received through the interface are put here and processed on this thread. Pairs of
+     * (docId, changeType).
+     */
+    private final BlockingQueue<Pair<String, Type>> changedDocIdQueue;
+
+    private long lastChangeTime = System.currentTimeMillis();
+    private Pair<String, Type> lastChangePair;
+
+    /**
+     * Creates a database change listener. Registers to the {@link CouchDbChangeMonitor} and fires events when the
+     * change monitor notifies that a document has changed.
+     *
+     * @param couchDb The CouchDB client to acces the database.
+     * @throws NunaliitException If an exception occurs while registering to the CouchDB change monitor.
+     */
+    public AbstractCouchDbChangeListener(CouchDb couchDb) throws NunaliitException {
+        super();
+        changedDocIdQueue = new LinkedBlockingQueue<>();
+        try {
+            CouchDbChangeMonitor changeMonitor = couchDb.getChangeMonitor();
+            changeMonitor.addChangeListener(this);
+        }
+        catch (Exception e) {
+            throw new NunaliitException("Problem registering to CouchDB change monitor", e);
+        }
+    }
+
+    @Override
+    public void run() {
+        running.set(true);
+        logger.info("Starting DB change listener thread");
+        performStartupTasks();
+
+        try {
+            while (running.get()) {
+                try {
+                    Pair<String, Type> docChanged = changedDocIdQueue.take();
+                    processDocIdChanged(docChanged);
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    logger.warn("DB change listener thread was interrupted. May have failed to complete an operation.");
+                }
+            }
+        }
+        finally {
+            logger.info("Shutting down DB change listener thread");
+        }
+    }
+
+    public void shutdown() {
+        running.set(false);
+    }
+
+    /**
+     * {@inheritDoc}
+     * Don't want to do any work in this method since it's run in the thread that calls it
+     * ({@link ca.carleton.gcrc.couch.client.impl.CouchDbChangeMonitorThread}). So we put the doc Id on a queue and
+     * process them in this thread's run method.
+     */
+    @Override
+    public void change(Type type, String docId, String rev, JSONObject rawChange, JSONObject doc) {
+        try {
```

Comment:
I realize there is only one working thread for AbstractCouchDbListener or its subclass, but manually synchronized this public interface seems to be a cautionary move
```suggestion
synchronized(this){
        try {
```

Reply:
I hear ya Decheng!  However, the only data that needs protecting is changedDocIdQueue, which is a thread-safe structure.  So there's no need to protect the entire public method.  That would be overkill and potentially degrade performance, unnecessarily.

Reply:
The changedDocIdQueue is thread-safe, but the lastChangePair can be access by different thread, which can potentially cause race condition issue

Reply:
Agreed.  Fixed.